src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,340 @@
+use std::io::Write;
+use std::path::Path;
+
+use anyhow::{anyhow, bail, Context};
+use candid::Principal;
+use clap::Parser;
+use fn_error_context::context;
+use ic_agent::{Agent, AgentError};
+use ic_wasm::metadata::get_metadata;
+use sha2::{Digest, Sha256};
+use slog::{error, info, trace, warn, Logger};
+
+/// Pull canisters upon which the project depends
+#[derive(Parser)]
+pub struct DepsPullOpts {}
+
+pub async fn exec(env: &dyn Environment, _opts: DepsPullOpts) -> DfxResult {
+    let logger = env.get_logger();
+    let project_root = env.get_config_or_anyhow()?.get_project_root().to_path_buf();
+
+    fetch_root_key_if_needed(env).await?;
+
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    let pull_canisters_in_config = get_pull_canisters_in_config(env)?;
+
+    let mut canisters_to_resolve: VecDeque<Principal> =
+        pull_canisters_in_config.values().cloned().collect();
+
+    let mut pulled_json = PulledJson::with_named(&pull_canisters_in_config);
+
+    while let Some(callee_canister) = canisters_to_resolve.pop_front() {
+        if !pulled_json.canisters.contains_key(&callee_canister) {
+            fetch_deps_to_pull(
+                agent,
+                logger,
+                callee_canister,
+                &mut canisters_to_resolve,
+                &mut pulled_json,
+            )
+            .await?;
+        }
+    }
+
+    let mut message = String::new();
+    message.push_str(&format!(
+        "Found {} dependencies:",
+        pulled_json.num_of_canisters()
+    ));
+    for id in pulled_json.get_all_ids() {
+        message.push('\n');
+        message.push_str(&id.to_text());
+    }
+
+    info!(logger, "{}", message);
+
+    let mut any_download_fail = false;
+
+    for (canister_id, pulled_canister) in pulled_json.canisters.iter_mut() {
+        if let Err(e) = download_canister_files(env, logger, *canister_id, pulled_canister).await {
+            error!(logger, "Failed to pull canister {canister_id}.\n{e}");
+            any_download_fail = true;
+        }
+    }
+
+    match any_download_fail {
+        true => bail!("Failed when pulling canisters."),
+        false => {
+            for (name, canister_id) in &pull_canisters_in_config {
+                copy_service_candid_to_project(&project_root, name, *canister_id)?;
+            }
+        }
+    }
+
+    save_pulled_json(&project_root, &pulled_json)?;
+    Ok(())
+}
+
+#[context("Failed to fetch and parse `dfx:deps` metadata from canister {canister_id}.")]
+async fn fetch_deps_to_pull(
+    agent: &Agent,
+    logger: &Logger,
+    canister_id: Principal,
+    canisters_to_pull: &mut VecDeque<Principal>,
+    pulled_json: &mut PulledJson,
+) -> DfxResult {
+    info!(
+        logger,
+        "Resolving dependencies of canister {canister_id}..."
+    );
+
+    match fetch_metatdata(agent, canister_id, DFX_DEPS).await {
+        Ok(Some(deps_raw)) => {
```

Comment:
Typo, and no need to match on the Result
```suggestion
    match fetch_metadata(agent, canister_id, DFX_DEPS).await? {
        Some(deps_raw) => {
```